rfc/2025-07-22-conditional-enable.md

```diff
@@ -0,0 +1,195 @@
+# Conditional field for enabling resources and modules
+
+## Current status
+
+Issue: https://github.com/opentofu/opentofu/issues/1306
+
+> [!NOTE]
+> Every time we refer to *resources* in this RFC, we're talking about [managed resources](https://opentofu.org/docs/language/resources/), [data sources](https://opentofu.org/docs/language/data-sources/) and [ephemeral resources](https://github.com/opentofu/opentofu/issues/2834).
+
+Right now, OpenTofu supports conditional enable/disable of resources by using a workaround with `count`.
+This approach brings a few problems, such as adding indexes to resources that would be single instances, making it harder to manage using these indexes.
+
+This RFC proposes a new way to do this in a cleaner and more semantic manner by using a new field on the `lifecycle` block called `enabled`:
+
+```
+resource "aws_instance" "example" {
+  # ...
+
+  lifecycle {
+    enabled = var.enable_server
+  }
+}
+```
+
+This is available not only for resources, but for modules too:
+
+```
+module "modcall" {
+  source = "./mod"
+
+  lifecycle {
+    enabled = var.env == "production"
+  }
+}
+```
+
+## Proposed Solution
+
+1. Raise errors if a resource is used while being disabled and the dependent resources are not disabled too;
+2. Support for conditional enabling on:
+  1. Resources - Add a new field on the `lifecycle` block called `enabled`;
+  2. Modules - Add a new `lifecycle` block for modules but only supporting the new `enabled` field.
+
+
+## Technical Approach
+
+### `lifecycle` support on Modules
+
+The lifecycle block is currently not supported on `modules`, but in our docs we mention to users that it's reserved for future usage:
+
+> OpenTofu does not use the lifecycle argument. However, the lifecycle block is reserved for future versions.
+
+https://opentofu.org/docs/language/modules/syntax/#meta-arguments
+
+It would be the right timing to start supporting it, but only for the `enabled` field.
+
+## Migration from existing resources
+
+### Migration from `count` to `enabled`
+
+As mentioned before, users were adding support for enabling/disabling resources by using `count = var.enabled ? 1 : 0` and this is the most common case where they want to use an `enabled` field.
+Luckily, this support is [in-place already](https://github.com/opentofu/opentofu/pull/3066#discussion_r2231518392). When using a `count` for this conditional dependency, if
```

Comment:
>Luckily, this support is [in-place already](https://github.com/opentofu/opentofu/pull/3066#discussion_r2231518392).

I feel like it took me a long time to realise that you were talking about moving resources here! a little bit of a pre-amble to each section would be appreciated to provide context of what we're talking about aside from the title.

For example, I think we should be saying something similar to: 

```
To support moving between resources where the `count = 0` workaround was used to the new `lifecycle -> enabled` approach, opentofu will automatically support an implicit move of the resource out of the box with no changes needed
```

What we have here is very jarring to read because i had no clue what "this support" is discussing and the linked comment was quite confusing.

Reply:
I think we should also note that this automatic behavior exists only for resources, and that for modules (unless we change something as part of implementing this feature) it will be necessary to write an explicit `moved` block:

```hcl
moved {
  from = module.example[0]
  to   = module.example
}
```

The current behavior of automatically generating an implied `moved` block for this situation only works for resources.